multi tenancy aware sequences

ThomasFOC opened the pull request:
Hey Rob,
the `TestSequenceMultiTenant` shows the problem with multi tenancy. To address this problem we implemented a delegator in the `BeanDescriptorManager`.

There is an older attempt of @rPraml where he uses `TenantDataSourceProvider` and `CurrentTenantProvider` and extends them to the `SequenceIdGenerator`:
https://github.com/FOCONIS/ebean/commit/2de1f215bbb164eef21b3f5cbe9f3d8475a78353

Do you have any other suggestions or is this simple attempt sufficient?

ThomasFOC committed:
failing test for sequences in multi tenant

This test case generates following stack:
```java
java.lang.NullPointerException
 at io.ebean.config.dbplatform.SequenceIdGenerator.getMoreIds(SequenceIdGenerator.java:158)
 at io.ebean.config.dbplatform.SequenceIdGenerator.loadMore(SequenceIdGenerator.java:117)
 at io.ebean.config.dbplatform.SequenceIdGenerator.nextId(SequenceIdGenerator.java:100)
 at io.ebeaninternal.server.deploy.BeanDescriptor.nextId(BeanDescriptor.java:1541)
 at io.ebeaninternal.server.core.DefaultServer.nextId(DefaultServer.java:851)
 at org.tests.idkeys.TestSequenceMultiTenant.test_multi_tenant_sequences(TestSequenceMultiTenant.java:35)
...
```

ThomasFOC committed:
Using DataSourceSupplier in BeanDescriptorManager#createSequenceIdGenerator instead of DataSource

To support multi tenancy for sequences a delegator is implemented in the BeanDescriptorManager

rPraml commented on ebean-core/src/main/java/io/ebeaninternal/server/deploy/BeanDescriptorManager.java:
```diff
@@ -1294,7 +1297,39 @@ private void checkMappedByManyToMany(DeployBeanPropertyAssocMany<?> prop) {
+      }
+
+      private PlatformIdGenerator get() {
+        return map.computeIfAbsent(dataSourceSupplier.getDataSource(), k -> create());
```
FI: This works good for a DS-supplier that supplies different dataSources for different tenants (As it is done in TenantMode.DB/DB_WITH_MASTER - which is our use case @foconis) - for other tenant modes like SCHEMA or CATALOG it may cause other problems. Maybe in a schema/catalog environment this should somehow configurable, where the Sequence lives.

rbygrave commented:
Hmmm, my initial thought is that this looks problematic.

ThomasFOC commented:
> Hmmm, my initial thought is that this looks problematic.

Hello @rbygrave ,
we tried to provide a test case for sequences in `TenantMode.SCHEMA`:
https://github.com/FOCONIS/ebean/commit/109958a1339f76f8dc336fc2ba29d7ad6ea78c45

We have tried to generate the schemas, when there is a `UserContext`-switch and provide the new schema.

An exception is raised in `io.ebean.config.dbplatform.SequenceIdGenerator.getMoreIds(int)` in line 158. To be more precise the exception comes from `io.ebeaninternal.server.core.MultiTenantDbSchemaSupplier.SchemaDataSource.getConnection()` where the schema is set for the connection.

This seems correct for us and may address the issue @rPraml sees.

But to dive deeper into this topic the test has to run over the problem that no schema seems to be generated.

After multiple hours of debugging and searching we are at our wits' end. Do you have an idea how to test `TenantMode.SCHEMA` with multiple schemas?

ThomasFOC committed:
Merge remote-tracking branch 'origin/master' into feature/multi-tenancy-aware-sequences

ThomasFOC committed:
failing test for tenantmode-schema-configuration

ThomasFOC commented on ebean-test/src/test/java/org/tests/idkeys/TestSequenceMultiTenant.java:
```diff
@@ -66,32 +68,100 @@ public DataSource dataSource(Object tenantId) {
   }
 
-  private static DataSource createDataSource(Object tenantId) {
+  private Database setupSchema() {
```
With the support of @rPraml we were able to write a failing test for tenantmode schema. THe main question seems to be, were the sequence schould live.

ThomasFOC committed:
Merge remote-tracking branch 'origin/master' into feature/multi-tenancy-aware-sequences

ThomasFOC committed:
compilation errors fixed

ThomasFOC committed:
fixed test setup and debugging information

should
```java
UserContext.set("5711", "2");
assertEquals(1L, db.nextId(GenKeySeqA.class));
```

or
```java
UserContext.set("5711", "2");
assertEquals(3L, db.nextId(GenKeySeqA.class));
```

Does the sequence live in tenant schema or globally?

ThomasFOC committed:
Merge remote-tracking branch 'origin/master' into feature/multi-tenancy-aware-sequences

ThomasFOC committed:
apply changes to test due restructuring of tests

rPraml commented:
Statusupdate: We know, that sequences in a multi-tenant environment are problematic as the sequence may be stored in the wrong database. We do NOT use them. This PR is not part of our fork.

@rbygrave I think the problem still exists... If you decide to close this PR, then it should be mentioned as limitation in multi-tenant setups with one DB/schema per tenant.

rPraml commented:
Hi @rbygrave  we will not pursue this topic any further as we do not need multi-tenant sequences.
You can decide, to close this issue (and maybe add a note to doc about known limitations on sequences + multi-tenant support)

rbygrave commented:
Replaced by https://github.com/ebean-orm/ebean/pull/3805
